TestingLowerBounds/ForMathlib/LeftRightDeriv.lean

```diff
@@ -0,0 +1,190 @@
+  exact rightDeriv_mono (ConvexOn.comp_neg hfc)
+
+lemma leftDeriv_le_rightDeriv : leftDeriv f ≤ rightDeriv f := by
+  intro x
+  rw [rightDeriv_eq_sInf_slope_of_convexOn hfc, leftDeriv_eq_sSup_slope_of_convexOn hfc]
+  refine csSup_le nonempty_of_nonempty_subtype ?_
+  rintro _ ⟨z, (zx : z < x), rfl⟩
+  refine le_csInf nonempty_of_nonempty_subtype ?_
+  rintro _ ⟨y, xy, rfl⟩
+  exact slope_mono hfc trivial ⟨trivial, zx.ne⟩ ⟨trivial, (Ne.symm (ne_of_lt xy))⟩ (zx.trans xy).le
+
+lemma rightDeriv_right_continuous_of_convexOn (w : ℝ) :
+    ContinuousWithinAt (rightDeriv f) (Ici w) w := by
+  simp_rw [← continuousWithinAt_Ioi_iff_Ici, ContinuousWithinAt]
+  have h_lim := MonotoneOn.tendsto_nhdsWithin_Ioi ((rightDeriv_mono hfc).monotoneOn (Ioi w))
+    (Monotone.map_bddBelow (rightDeriv_mono hfc) bddBelow_Ioi)
+  set l := sInf (rightDeriv f '' Ioi w)
+  convert h_lim
+  refine (LE.le.le_iff_eq ?_).mp ?_ --any better way to split an equality goal into the two inequalitites?
+  · rw [rightDeriv_eq_sInf_slope_of_convexOn hfc]
+    refine le_csInf nonempty_of_nonempty_subtype ?_ --is there any way to avoid the rintro here? if I just use fun inside the refine it does not work, it seems that the rfl inside the pattern is not supported by the refine tactic
+    rintro _ ⟨y, (wy : w < y), rfl⟩
+    have slope_lim : Tendsto (slope f y) (𝓝[>] w) (𝓝 (slope f y w)) := by
+      have hf_cont : ContinuousWithinAt f (Ioi w) w := -- I would like to replace this with a lemma that derives the continuity from the convexity, it seems that this result is still not in mathlib, see https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Continuity.20.20of.20convex.20functions, they are in the process of proving it in the LeanCamCombi project
+        DifferentiableWithinAt.continuousWithinAt (differentiableWithinAt_Ioi_of_convexOn hfc w)
+      exact ((continuousWithinAt_id.sub continuousWithinAt_const).inv₀
+        (sub_ne_zero.2 <| ne_of_lt wy)).smul
+        (hf_cont.sub continuousWithinAt_const) |>.tendsto
+    rw [slope_comm] at slope_lim
+    refine le_of_tendsto_of_tendsto h_lim slope_lim ?_
+    rw [← nhdsWithin_Ioo_eq_nhdsWithin_Ioi wy]
+    refine eventually_nhdsWithin_of_forall fun z hz ↦ ?_
+    rw [slope_comm, rightDeriv_eq_sInf_slope_of_convexOn hfc]
+    exact csInf_le (bddBelow_slope_Ioi_of_convexOn hfc z) ⟨y, hz.2, rfl⟩
+  · exact ge_of_tendsto h_lim <| eventually_nhdsWithin_of_forall
+      fun y hy ↦ rightDeriv_mono hfc (le_of_lt hy)
+
+lemma leftDeriv_left_continuous_convexOn (w : ℝ) : ContinuousWithinAt (leftDeriv f) (Iic w) w := by
+  have h_map : MapsTo Neg.neg (Iic w) (Ici (-w)) := fun _ (h : _ ≤ w) ↦ (neg_le_neg_iff.mpr h)
+  rw [leftDeriv_eq_rightDeriv]
+  refine ContinuousWithinAt.comp ?_ continuousWithinAt_neg h_map |>.neg
+  exact rightDeriv_right_continuous_of_convexOn hfc.comp_neg (-w)
+
+end ConvexOn
+
+noncomputable
+def StieltjesFunction.rightDeriv_of_convex (f : ℝ → ℝ) (hf : ConvexOn ℝ univ f) :
```

Comment:
Perhaps call this `ConvexOn.rightDerivStieltjes` to enable dot notation on the `ConvexOn` hypothesis?